File: clients/fsharp/Protocol.fs
namespace NetworkMutex

open System
open System.Collections.Generic
open System.Text.Json

type RequestType =
  | Version
  | Auth
  | Lock
  | Unlock
  | RegisterRead
  | RegisterWrite
  | EndRead
  | EndWrite
  | LockInfo
  | Ls
  | Heartbeat

type ResponseType =
  | Version
  | Auth
  | Lock
  | CompositeLock
  | Unlock
  | RegisterReadResult
  | RegisterWriteResult
  | EndReadResult
  | EndWriteResult
  | LockInfo
  | LsResult
  | Reelection
  | Error
  | Ok
  | Unknown

type Response =
  { Type: ResponseType
    Uuid: string
    Key: string option
    Keys: string list
    Acquired: bool option
    Unlocked: bool option
    Granted: bool option
    LockUuid: string option
    FencingToken: uint64 option
    FencingTokens: Map<string, uint64>
    Error: string option }

module Protocol =
  [<Literal>]
  let ProtocolVersion = "0.1.0"

  [<Literal>]
  let MaxCompositeKeys = 5

  [<Literal>]
  let MaxFencingToken = 9_007_199_254_740_991UL

  let requestTypes =
    [ "version"
      "auth"
      "lock"
      "unlock"
      "registerRead"
      "registerWrite"
      "endRead"
      "endWrite"
      "lockInfo"
      "ls"
      "heartbeat" ]

  let responseTypes =
    [ "version"
      "auth"
      "lock"
      "compositeLock"
      "unlock"
      "registerReadResult"
      "registerWriteResult"
      "endReadResult"
      "endWriteResult"
      "lockInfo"
      "lsResult"
      "reelection"
      "error"
      "ok" ]

  let requestTypeToWire =
    function
    | RequestType.Version -> "version"
    | RequestType.Auth -> "auth"
    | RequestType.Lock -> "lock"
    | RequestType.Unlock -> "unlock"
    | RequestType.RegisterRead -> "registerRead"
    | RequestType.RegisterWrite -> "registerWrite"
    | RequestType.EndRead -> "endRead"
    | RequestType.EndWrite -> "endWrite"
    | RequestType.LockInfo -> "lockInfo"
    | RequestType.Ls -> "ls"
    | RequestType.Heartbeat -> "heartbeat"

  let responseTypeFromWire =
    function
    | "version" -> ResponseType.Version
    | "auth" -> ResponseType.Auth
    | "lock" -> ResponseType.Lock
    | "compositeLock" -> ResponseType.CompositeLock
    | "unlock" -> ResponseType.Unlock
    | "registerReadResult" -> ResponseType.RegisterReadResult
    | "registerWriteResult" -> ResponseType.RegisterWriteResult
    | "endReadResult" -> ResponseType.EndReadResult
    | "endWriteResult" -> ResponseType.EndWriteResult
    | "lockInfo" -> ResponseType.LockInfo
    | "lsResult" -> ResponseType.LsResult
    | "reelection" -> ResponseType.Reelection
    | "error" -> ResponseType.Error
    | "ok" -> ResponseType.Ok
    | _ -> ResponseType.Unknown

  let private frame (fields: (string * obj option) list) =
    let dict = Dictionary<string, obj>()

    for key, value in fields do
      match value with
      | Some v -> dict[key] <- v
      | None -> ()

    JsonSerializer.Serialize(dict) + "\n"

  let versionRequest uuid value =
    frame [ "type", Some(box "version"); "uuid", Some(box uuid); "value", Some(box value) ]

  let authRequest uuid token =
    frame [ "type", Some(box "auth"); "uuid", Some(box uuid); "token", Some(box token) ]

  let lockRequestSingle uuid key ttlMs maxHolders wait =
    frame
      [ "type", Some(box "lock")
        "uuid", Some(box uuid)
        "key", Some(box key)
        "ttl", (if ttlMs > 0 then Some(box ttlMs) else None)
        "max", Option.map box maxHolders
        "wait", Option.map box wait ]

  let lockRequestComposite uuid keys ttlMs wait =
    let count = List.length keys

    if count < 1 || count > MaxCompositeKeys then
      invalidArg (nameof keys) $"composite key count must be 1..=5, got {count}"

    frame
      [ "type", Some(box "lock")
        "uuid", Some(box uuid)
        "keys", Some(box (keys |> List.toArray))
        "ttl", (if ttlMs > 0 then Some(box ttlMs) else None)
        "wait", Option.map box wait ]

  let unlockRequestSingle uuid key lockUuid force =
    frame
      [ "type", Some(box "unlock")
        "uuid", Some(box uuid)
        "key", Some(box key)
        "lockUuid", (lockUuid |> Option.filter (fun s -> s <> "") |> Option.map box)
        "force", (if force then Some(box true) else None) ]

  let unlockRequestComposite uuid keys lockUuid =
    frame
      [ "type", Some(box "unlock")
        "uuid", Some(box uuid)
        "keys", Some(box (keys |> List.toArray))
        "lockUuid", (lockUuid |> Option.filter (fun s -> s <> "") |> Option.map box) ]

  let rwRequest requestType uuid key =
    frame [ "type", Some(box (requestTypeToWire requestType)); "uuid", Some(box uuid); "key", Some(box key) ]

  let lockInfoRequest uuid key = rwRequest RequestType.LockInfo uuid key
  let lsRequest uuid = frame [ "type", Some(box "ls"); "uuid", Some(box uuid) ]
  let heartbeatRequest uuid = frame [ "type", Some(box "heartbeat"); "uuid", Some(box uuid) ]

  let private tryGetString (name: string) (root: JsonElement) =
    match root.TryGetProperty name with
    | true, value when value.ValueKind = JsonValueKind.String -> Some(value.GetString())
    | _ -> None

  let private tryGetBool (name: string) (root: JsonElement) =
    match root.TryGetProperty name with
    | true, value when value.ValueKind = JsonValueKind.True || value.ValueKind = JsonValueKind.False ->
      Some(value.GetBoolean())
    | _ -> None

  let private tryGetUInt64 (name: string) (root: JsonElement) =
    match root.TryGetProperty name with
    | true, value when value.ValueKind = JsonValueKind.Number ->
      match value.TryGetUInt64() with
      | true, n -> Some n
      | _ -> None
    | _ -> None

  let private getStringList (name: string) (root: JsonElement) =
    match root.TryGetProperty name with
    | true, value when value.ValueKind = JsonValueKind.Array ->
      value.EnumerateArray()
      |> Seq.map (fun item -> item.GetString())
      |> Seq.filter (fun item -> not (isNull item))
      |> Seq.toList
    | _ -> []

  let private getUInt64Map (name: string) (root: JsonElement) =
    match root.TryGetProperty name with
    | true, value when value.ValueKind = JsonValueKind.Object ->
      value.EnumerateObject()
      |> Seq.choose (fun prop ->
        match prop.Value.TryGetUInt64() with
        | true, n -> Some(prop.Name, n)
        | _ -> None)
      |> Map.ofSeq
    | _ -> Map.empty

  let private validToken =
    function
    | Some token when token >= 1UL && token <= MaxFencingToken -> true
    | _ -> false

  let private validTokenValue token =
    token >= 1UL && token <= MaxFencingToken

  let private nonEmpty =
    function
    | Some value when not (String.IsNullOrWhiteSpace value) -> true
    | _ -> false

  let private validateAuthority response =
    match response.Type, response.Acquired, response.Granted with
    | ResponseType.Lock, Some true, _ ->
      if not (nonEmpty response.Key && nonEmpty response.LockUuid && validToken response.FencingToken) then
        raise (JsonException("successful lock grant omitted valid fenced authority"))
    | ResponseType.CompositeLock, Some true, _ ->
      let distinctKeys = response.Keys |> Set.ofList |> Set.count
      let completeTokens =
        response.Keys
        |> List.forall (fun key ->
          match Map.tryFind key response.FencingTokens with
          | Some token -> validTokenValue token
          | None -> false)

      if not (nonEmpty response.LockUuid
              && response.Keys.Length >= 1
              && response.Keys.Length <= MaxCompositeKeys
              && distinctKeys = response.Keys.Length
              && Map.count response.FencingTokens = response.Keys.Length
              && completeTokens) then
        raise (JsonException("successful composite grant omitted complete fenced authority"))
    | (ResponseType.RegisterReadResult | ResponseType.RegisterWriteResult), _, Some true ->
      if not (nonEmpty response.Key && nonEmpty response.LockUuid && validToken response.FencingToken) then
        raise (JsonException("successful reader/writer grant omitted valid fenced authority"))
    | _ -> ()

    response

  let decodeResponse (line: string) =
    use doc = JsonDocument.Parse(line)
    let root = doc.RootElement

    let response =
      { Type = responseTypeFromWire (tryGetString "type" root |> Option.defaultValue "")
        Uuid = tryGetString "uuid" root |> Option.defaultValue ""
        Key = tryGetString "key" root
        Keys = getStringList "keys" root
        Acquired = tryGetBool "acquired" root
        Unlocked = tryGetBool "unlocked" root
        Granted = tryGetBool "granted" root
        LockUuid = tryGetString "lockUuid" root
        FencingToken = tryGetUInt64 "fencingToken" root
        FencingTokens = getUInt64Map "fencingTokens" root
        Error = tryGetString "error" root }

    validateAuthority response

File: clients/csharp/Protocol.cs
using System.Text.Json;

namespace NetworkMutex;

public enum RequestType
{
    Version,
    Auth,
    Lock,
    Unlock,
    RegisterRead,
    RegisterWrite,
    EndRead,
    EndWrite,
    LockInfo,
    Ls,
    Heartbeat,
}

public enum ResponseType
{
    Version,
    Auth,
    Lock,
    CompositeLock,
    Unlock,
    RegisterReadResult,
    RegisterWriteResult,
    EndReadResult,
    EndWriteResult,
    LockInfo,
    LsResult,
    Reelection,
    Error,
    Ok,
    Unknown,
}

public sealed record Response(
    ResponseType Type,
    string Uuid,
    string? Key,
    IReadOnlyList<string> Keys,
    bool? Acquired,
    bool? Unlocked,
    string? LockUuid,
    ulong? FencingToken,
    IReadOnlyDictionary<string, ulong> FencingTokens,
    string? Error);

public static class Protocol
{
    public const string ProtocolVersion = "0.1.0";
    public const int MaxCompositeKeys = 5;

    public static readonly string[] RequestTypes =
    [
        "version",
        "auth",
        "lock",
        "unlock",
        "registerRead",
        "registerWrite",
        "endRead",
        "endWrite",
        "lockInfo",
        "ls",
        "heartbeat",
    ];

    public static readonly string[] ResponseTypes =
    [
        "version",
        "auth",
        "lock",
        "compositeLock",
        "unlock",
        "registerReadResult",
        "registerWriteResult",
        "endReadResult",
        "endWriteResult",
        "lockInfo",
        "lsResult",
        "reelection",
        "error",
        "ok",
    ];

    public static string ToWire(RequestType type) => type switch
    {
        RequestType.Version => "version",
        RequestType.Auth => "auth",
        RequestType.Lock => "lock",
        RequestType.Unlock => "unlock",
        RequestType.RegisterRead => "registerRead",
        RequestType.RegisterWrite => "registerWrite",
        RequestType.EndRead => "endRead",
        RequestType.EndWrite => "endWrite",
        RequestType.LockInfo => "lockInfo",
        RequestType.Ls => "ls",
        RequestType.Heartbeat => "heartbeat",
        _ => throw new ArgumentOutOfRangeException(nameof(type), type, null),
    };

    public static ResponseType ResponseTypeFromWire(string value) => value switch
    {
        "version" => ResponseType.Version,
        "auth" => ResponseType.Auth,
        "lock" => ResponseType.Lock,
        "compositeLock" => ResponseType.CompositeLock,
        "unlock" => ResponseType.Unlock,
        "registerReadResult" => ResponseType.RegisterReadResult,
        "registerWriteResult" => ResponseType.RegisterWriteResult,
        "endReadResult" => ResponseType.EndReadResult,
        "endWriteResult" => ResponseType.EndWriteResult,
        "lockInfo" => ResponseType.LockInfo,
        "lsResult" => ResponseType.LsResult,
        "reelection" => ResponseType.Reelection,
        "error" => ResponseType.Error,
        "ok" => ResponseType.Ok,
        _ => ResponseType.Unknown,
    };

    public static string VersionRequest(string uuid, string value = ProtocolVersion) =>
        Frame(new Dictionary<string, object?> { ["type"] = "version", ["uuid"] = uuid, ["value"] = value });

    public static string AuthRequest(string uuid, string token) =>
        Frame(new Dictionary<string, object?> { ["type"] = "auth", ["uuid"] = uuid, ["token"] = token });

    public static string LockRequestSingle(
        string uuid,
        string key,
        long ttlMs = 0,
        int? maxHolders = null,
        bool? wait = null) =>
        Frame(new Dictionary<string, object?>
        {
            ["type"] = "lock",
            ["uuid"] = uuid,
            ["key"] = key,
            ["ttl"] = ttlMs > 0 ? ttlMs : null,
            ["max"] = maxHolders,
            ["wait"] = wait,
        });

    public static string LockRequestComposite(string uuid, IReadOnlyList<string> keys, long ttlMs = 0, bool? wait = null)
    {
        if (keys.Count is < 1 or > MaxCompositeKeys)
        {
            throw new ArgumentException($"composite key count must be 1..=5, got {keys.Count}", nameof(keys));
        }

        return Frame(new Dictionary<string, object?>
        {
            ["type"] = "lock",
            ["uuid"] = uuid,
            ["keys"] = keys,
            ["ttl"] = ttlMs > 0 ? ttlMs : null,
            ["wait"] = wait,
        });
    }

    public static string UnlockRequestSingle(string uuid, string key, string? lockUuid, bool force = false) =>
        Frame(new Dictionary<string, object?>
        {
            ["type"] = "unlock",
            ["uuid"] = uuid,
            ["key"] = key,
            ["lockUuid"] = string.IsNullOrEmpty(lockUuid) ? null : lockUuid,
            ["force"] = force ? true : null,
        });

    public static string UnlockRequestComposite(string uuid, IReadOnlyList<string> keys, string? lockUuid) =>
        Frame(new Dictionary<string, object?>
        {
            ["type"] = "unlock",
            ["uuid"] = uuid,
            ["keys"] = keys,
            ["lockUuid"] = string.IsNullOrEmpty(lockUuid) ? null : lockUuid,
        });

    public static string RwRequest(RequestType type, string uuid, string key) =>
        Frame(new Dictionary<string, object?> { ["type"] = ToWire(type), ["uuid"] = uuid, ["key"] = key });

    public static string LockInfoRequest(string uuid, string key) => RwRequest(RequestType.LockInfo, uuid, key);

    public static string LsRequest(string uuid) =>
        Frame(new Dictionary<string, object?> { ["type"] = "ls", ["uuid"] = uuid });

    public static string HeartbeatRequest(string uuid) =>
        Frame(new Dictionary<string, object?> { ["type"] = "heartbeat", ["uuid"] = uuid });

    public static Response DecodeResponse(string jsonLine)
    {
        using var doc = JsonDocument.Parse(jsonLine);
        var root = doc.RootElement;
        var type = ResponseTypeFromWire(GetString(root, "type") ?? "");
        var keys = GetStringArray(root, "keys");
        var fencingTokens = GetUInt64Map(root, "fencingTokens");

        return new Response(
            type,
            GetString(root, "uuid") ?? "",
            GetString(root, "key"),
            keys,
            GetBool(root, "acquired"),
            GetBool(root, "unlocked"),
            GetString(root, "lockUuid"),
            GetUInt64(root, "fencingToken"),
            fencingTokens,
            GetString(root, "error"));
    }

    private static string Frame(IDictionary<string, object?> fields)
    {
        var compact = fields.Where(kv => kv.Value is not null).ToDictionary(kv => kv.Key, kv => kv.Value);
        return JsonSerializer.Serialize(compact) + "\n";
    }

    private static string? GetString(JsonElement root, string name) =>
        root.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.String ? value.GetString() : null;

    private static bool? GetBool(JsonElement root, string name) =>
        root.TryGetProperty(name, out var value) && value.ValueKind is JsonValueKind.True or JsonValueKind.False
            ? value.GetBoolean()
            : null;

    private static ulong? GetUInt64(JsonElement root, string name) =>
        root.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.Number && value.TryGetUInt64(out var n)
            ? n
            : null;

    private static IReadOnlyList<string> GetStringArray(JsonElement root, string name)
    {
        if (!root.TryGetProperty(name, out var value) || value.ValueKind != JsonValueKind.Array)
        {
            return Array.Empty<string>();
        }

        return value.EnumerateArray().Select(item => item.GetString() ?? "").ToArray();
    }

    private static IReadOnlyDictionary<string, ulong> GetUInt64Map(JsonElement root, string name)
    {
        if (!root.TryGetProperty(name, out var value) || value.ValueKind != JsonValueKind.Object)
        {
            return new Dictionary<string, ulong>();
        }

        var map = new Dictionary<string, ulong>();
        foreach (var prop in value.EnumerateObject())
        {
            if (prop.Value.ValueKind == JsonValueKind.Number && prop.Value.TryGetUInt64(out var n))
            {
                map[prop.Name] = n;
            }
        }

        return map;
    }
}

